### PR TITLE
Add a MongoDB client extension

### PR DESCRIPTION
**Pull request by cescoffier**
This is the first version of the Mongo extension.

It:

* works in both JVM and native mode
* proposes an imperative client (regular mongo client) and a reactive one (based on the reactive streams mongo driver)
* supports authentication and TLS (even in native mode), but does not support `mongo+srv://` in native mode (SVM limitation)
* supports the hot reload

### CODE REVIEW

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
== Running a MongoDB Database
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
As by default, `MongoClient` is configure to access a local MongoDB database on port 27017 (the default MongoDB port), if you have a local running database on this port, there is nothing more to do before being able to test it!
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
If you want to use Docker to run a MongoDB database, you can use the following command to launch one:
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]: I don't think you're referring to the right quickstart?

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
== Reactive MongoDB Client
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
A reactive MongoDB Client is included in Quarkus.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Using it is as easy as using the classic MongoDB Client. You can rewrite the previous example to use it like the following.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
To facilitate this there are two ways.
```
?

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]: Same comment about MongoDB but I suppose it depends on how you want to name the extension in the end.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a MongoDB `Document` automatically.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
First you need to create a Bson `Codec` that will tell Bson how to transform your entity to/from a MongoDB `Document`.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Here we use a `CollectibleCodec` as our object is retrievable from the database (it has a MongoDB identifier), if not we would have used a `Codec` instead.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
More information in the codec documentation: https://mongodb.github.io/mongo-java-driver/3.10/bson/codecs.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Then you need to create a `CodecProvider` to link this `Codec` to the `Fruit` class.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Accessing a MongoDB database from a Mongo Client is easy with Quarkus as it provides configuration and native support for it.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Comma separated list of hosts.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Minimum number of connections in the connection pool.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]: Should they be `Duration` instead? Same question for all these variables.

(we have a note to include if this is the case)

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]: maxPoolSi?

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
If connecting with TLS, this option enables insecure TLS connections. Defaults to false.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
Whether to connect using TLS. Defaults to false.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]:
```suggestion
server version. For the GSSAPI and MONGODB-X509 mechanisms, no password is accepted, only the username.
```

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/mongo-guide.adoc]: `the database that the credentials have been created` hmmm, `that` is weird here.

**Review by gsmet — CHANGES_REQUESTED** [review]
I made a big review of the documentation part, NOT the code.

I think we should try to be consistent about the Mongo*DB* naming but I let you decide what you prefer.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: ze :-D

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java
Comment (cescoffier) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java]: Agreed!

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Here we cannot say MongoDB database, as DB is for database.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Same comment as above, when followed with database, it should be mongo.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: it's copying the same file, but yes, would be better to provide the file in the right quickstart. Keeping this comment open until we have the right url.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: section dropped.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Here as it's not followed with database, it should be MongoDB.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Wondering about it. This is the MongClient configuration passed almost directly to the configuration. But yes, I would prefer using duration.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Updated. This was actually in the upstream code...

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Sentence rewritten.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: fixed

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: fixed.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: fixed.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Also had to be fixed in the javadoc.

File: docs/src/main/asciidoc/mongo-guide.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/mongo-guide.adoc]: Fixed - used duration everywhere.

**Comment by cescoffier** [comment]
Ok, I believe we reached a milestone. I'm going to re-organized my commits and ask for a review.

File: extensions/mongo-client/deployment/pom.xml
Comment (gsmet) [inline extensions/mongo-client/deployment/pom.xml]:
```suggestion
    <name>Quarkus - MongoDB Client - Deployment</name>
```
would be better, I think.

File: extensions/mongo-client/pom.xml
Comment (gsmet) [inline extensions/mongo-client/pom.xml]:
```suggestion
   <name>Quarkus - MongoDB Client</name>
```

File: extensions/mongo-client/runtime/pom.xml
Comment (gsmet) [inline extensions/mongo-client/runtime/pom.xml]:
```suggestion
    <name>Quarkus - MongoDB Client - Runtime</name>
```

File: extensions/pom.xml
Comment (gsmet) [inline extensions/pom.xml]: It should be mongodb-client for consistency.

File: integration-tests/pom.xml
Comment (gsmet) [inline integration-tests/pom.xml]: It should be mongodb-client for consistency.

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java
Comment (gsmet) [inline extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java]: I think I would have made the package `io.quarkus.mongodb.client` (same for the ITs).

But I can understand how it's not practical to change it now. So feel free to ignore.

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java
Comment (gsmet) [inline extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java]: Not excited about using the `toString()`. People sometimes change that without thinking that it could be used.

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java
Comment (gsmet) [inline extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java]: Does your extension work if there's no config at all? We usually try to support that.

File: extensions/mongo-client/runtime/src/main/java/io/quarkus/mongo/runtime/MongoClientTemplate.java
Comment (gsmet) [inline extensions/mongo-client/runtime/src/main/java/io/quarkus/mongo/runtime/MongoClientTemplate.java]: You don't want to use that. Please use JBoss Logging.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added another round of comments.

I still think there are too many commits. Typically https://github.com/quarkusio/quarkus/pull/3022/commits/ac046f26c2551472862f8f1db82b567b412e192c should be squashed as it really has no value.

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java
Comment (gsmet) [inline extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java]: You don't need to test that, it's automatically taken care of, just add the build item.

File: extensions/mongo-client/deployment/pom.xml
Comment (cescoffier) [reply extensions/mongo-client/deployment/pom.xml]: I changed the artifactId, forgot about the name. So definitely!

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java
Comment (cescoffier) [reply extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java]: It needs to be done. I actually started, but forgot to commit and pushed. It will be done today.

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java
Comment (cescoffier) [reply extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java]: That's a good question, let me try ;-)

Comment (cescoffier) [reply extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java]: So, yes it works, and it connects to `mongodb://localhost:27017` (default host and port).

File: extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java
Comment (cescoffier) [reply extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java]: Done.

**Comment by cescoffier** [comment]
I've reduced the number of commits.

**Comment by gsmet** [comment]
@cescoffier could you rebase, there are some conflicts with the pom files. It should be simple enough to fix.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/ReactiveMongoClientImpl.java
Comment (gsmet) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/ReactiveMongoClientImpl.java]: The convention until now was to have the impl classes in the `runtime` package so it should be `io.quarkus.mongodb.runtime`.

Comment (cescoffier) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/ReactiveMongoClientImpl.java]: This module exposes an API, it would be weird to be in a `runtime` quarkus-specific structure. That being said, for implementation, I don't see an issue.

**Comment by cescoffier** [comment]
Rebase done.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/ReactiveMongoClientImpl.java
Comment (gsmet) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/ReactiveMongoClientImpl.java]: Well, this is the convention we decided. I wanted impl/spi packages but got ruled out.

So right now, the conventions we have are:
`io.quarkus.<extension>`: the potential runtime API
`io.quarkus.<extension>.runtime`: everything that is not API (you can have an `impl` subpackage here if you want)
`io.quarkus.<extension>.deployment`: everything from the deployment artifact

**Comment by gsmet** [comment]
I properly rebased to avoid the merge commit as it makes the history harder to read and adjusted a few things.

I will wait for CI and merge.

**Comment by rsvoboda** [comment]
Thanks, just noticed that quarkus-quickstarts are broken due to missing quarkus-mongodb-client dependency for using-mongodb-client qs

**Comment by gsmet** [comment]
@rsvoboda yes, I'm waiting for CI before merging this one.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
And merged! Thanks @cescoffier and @loicmathieu !

**Pull request merged**
